server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+
+	/**
+	 * Map of registered metric slugs and their registered data.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $registered_metrics_data = array();
+
+	/**
+	 * Registers a metric to calculate for the Server-Timing header.
+	 *
+	 * This method must be called before the {@see 'perflab_server_timing_send_header'} hook.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $metric_slug The metric slug.
+	 * @param array  $args        {
+	 *     Arguments for the metric.
+	 *
+	 *     @type callable $measure_callback The callback that initiates calculating the metric value. It will receive
+	 *                                      the Perflab_Server_Timing_Metric instance as a parameter, in order to set
+	 *                                      the value when it has been calculated. Metric values must be provided in
+	 *                                      milliseconds.
+	 *     @type string   $access_cap       Capability required to view the metric. If this is a public metric, this
+	 *                                      needs to be set to "exist".
+	 * }
+	 */
+	public function register_metric( $metric_slug, array $args ) {
+		if ( isset( $this->registered_metrics[ $metric_slug ] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: metric slug */
+				sprintf( __( 'A metric with the slug %s is already registered.', 'performance-lab' ), $metric_slug ),
+				''
+			);
+			return;
+		}
+
+		if ( did_action( 'perflab_server_timing_send_header' ) && ! doing_action( 'perflab_server_timing_send_header' ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: WordPress action name */
+				sprintf( __( 'The method must be called before or during the %s action.', 'performance-lab' ), 'perflab_server_timing_send_header' ),
+				''
+			);
+			return;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			array(
+				'measure_callback' => null,
+				'access_cap'       => null,
+			)
+		);
+		if ( ! $args['measure_callback'] || ! is_callable( $args['measure_callback'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a callable.', 'performance-lab' ), '$args["measure_callback"]' ),
+				''
+			);
+			return;
+		}
+		if ( ! $args['access_cap'] || ! is_string( $args['access_cap'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a string.', 'performance-lab' ), '$args["access_cap"]' ),
+				''
+			);
+			return;
+		}
+
+		$this->registered_metrics[ $metric_slug ]      = new Perflab_Server_Timing_Metric( $metric_slug );
+		$this->registered_metrics_data[ $metric_slug ] = $args;
+
+		// If the current user has already been determined and they lack the necessary access,
+		// do not even attempt to calculate the metric.
+		if ( did_action( 'set_current_user' ) && ! current_user_can( $args['access_cap'] ) ) {
+			return;
+		}
```

Comment:
Initially I was unsure about this, but I think @ankitrox _might_ have a point here, especially given that we instantiate the Server Timing API at object-cache.php drop-in level, which essentially allows the use of the Server Timing API from that point in the execution cycle. 

What I am unsure about though, is whether a user has been established at this point or not?

Reply:
> What if the current user is not determined? I think it will still run the callback?

Don't we want to gather metrics for non-authenticated users as well? authenticated loads are very different that anonymous loads.

Reply:
@ankitrox @10upsimon Yes, it will still be run. However, that does not present a real issue. Here's why:

* The primary purpose of the `access_cap` is to control whether the metric is exposed in the `Server-Timing` header.
* Because of that, the cap is checked when rendering the header (in the `get_header_value()` method). That is the actually critical check. And at that point, for sure the user will have been set.
* The check here is just a very minor performance consideration: Basically _if_ at this point the user is already set, we can check the capability right away and not even calculate the metric.
* If the user isn't set at that point, running the function to calculate the metric is perfectly fine; alternatively we _could_ ensure the current user is set early (normally it's set right before the `init` action), but that's not something I like to do especially given the little benefit it would bring here.

Let me know if that satisfies your concerns.

Reply:
@felixarntz this makes total sense to me now, thank you for clarifying.